Fix Source.actorRef not  completing for Status.Success("ok")

giftig opened the pull request:
sbt hanged when I tried to do a `test-only` on this locally so I'm going to let your CI check it.

This attempts to reproduce #25285

giftig committed:
Source.actorRef not completing on Success (#25285)

When a Success is received, call onCompleteThenStop instead of just
context.stop; that takes care of the completion logic instead of just
stopping the actor and leaving the stream going.

Add test to ensure the stream materializes on Source.actorRef receiving
Status.Success

Remove tests around stream completion behaviour in response to
PoisonPill - as well as these tests not correctly demonstrating that the
completion was passed on downstream, they describe behaviour which was
previously incidental and is no longer accurate.

Update the docs to reflect that PoisonPill should not be used on the
actor ref as this scenario will necessarily result in bad behaviour as
it will be unable to signal the completion downstream.

Make a few grammar fixes and remove some trailing space while updating the
docs.

lightbend-cla-validator commented:
Hi @giftig,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

giftig commented:
Ah, your CI needs approval. Okay, I'll try again to run it locally.

ktoso commented:
OK TO TEST

giftig commented:
I managed to build locally this time and received the same failure as I did in my own project, so hopefully once jenkins is finished you'll see it reproduced reliably :+1: 

In the meantime I'll work around it in my tests; the rest of your test suite shows me how to do that anyway.

```
[info] ActorRefSourceSpec:
Coroner Thread Count starts at 14 in akka.stream.scaladsl.ActorRefSourceSpec
[info] A ActorRefSource
[info] - must emit received messages to the stream (567 milliseconds)
[info] - must buffer when needed (22 milliseconds)
[info] - must drop new when full and with dropNew strategy (17 milliseconds)
[info] - must terminate when the stream is cancelled (38 milliseconds)
[info] - must not fail when 0 buffer space and demand is signalled (33 milliseconds)
[info] - must complete the stream immediatly when receiving PoisonPill (38 milliseconds)
[info] - must signal buffered elements and complete the stream after receiving Status.Success (38 milliseconds)
[info] - must signal buffered elements and complete the stream after receiving a Status.Success companion (34 milliseconds)
[info] - must not buffer elements after receiving Status.Success (38 milliseconds)
[WARN] [07/01/2018 10:28:57.200] [ActorRefSourceSpec-akka.actor.default-dispatcher-5] [akka://ActorRefSourceSpec/user/StreamSupervisor*] received dead letter from Actor[akka://ActorRefSourceSpec/system/testProbe-24#-1497055377]: GetChildren
--- Stream actors debug dump ---
Stream is completed. No debug information is available
[info] - must complete and materialize the stream after receiving Status.Success *** FAILED *** (7 seconds, 58 milliseconds)
[info]   A timeout occurred waiting for a future to complete. Queried 51 times, sleeping 100 milliseconds between each query. (ActorRefSourceSpec.scala:138)
[info]   org.scalatest.concurrent.Futures$FutureConcept$$anon$1:
[info]   at org.scalatest.concurrent.Futures$FutureConcept$class.tryTryAgain$1(Futures.scala:538)
[info]   at org.scalatest.concurrent.Futures$FutureConcept$class.futureValueImpl(Futures.scala:550)
[info]   at org.scalatest.concurrent.ScalaFutures$$anon$1.futureValueImpl(ScalaFutures.scala:275)
[info]   at org.scalatest.concurrent.Futures$FutureConcept$class.futureValue(Futures.scala:476)
[info]   at org.scalatest.concurrent.ScalaFutures$$anon$1.futureValue(ScalaFutures.scala:275)
[info]   at akka.stream.scaladsl.ActorRefSourceSpec$$anonfun$1$$anonfun$apply$mcV$sp$22$$anonfun$apply$7.apply(ActorRefSourceSpec.scala:138)
[info]   at akka.stream.scaladsl.ActorRefSourceSpec$$anonfun$1$$anonfun$apply$mcV$sp$22$$anonfun$apply$7.apply(ActorRefSourceSpec.scala:133)
[info]   at akka.stream.testkit.scaladsl.StreamTestKit$.assertAllStagesStopped(StreamTestKit.scala:28)
[info]   at akka.stream.scaladsl.ActorRefSourceSpec$$anonfun$1$$anonfun$apply$mcV$sp$22.apply(ActorRefSourceSpec.scala:133)
[info]   at akka.stream.scaladsl.ActorRefSourceSpec$$anonfun$1$$anonfun$apply$mcV$sp$22.apply(ActorRefSourceSpec.scala:133)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:1078)
[info]   at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
[info]   at akka.stream.testkit.StreamSpec.withFixture(StreamSpec.scala:27)
[info]   at org.scalatest.WordSpecLike$class.invokeWithFixture$1(WordSpecLike.scala:1075)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:1088)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:1088)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
[info]   at org.scalatest.WordSpecLike$class.runTest(WordSpecLike.scala:1088)
[info]   at akka.testkit.AkkaSpec.runTest(AkkaSpec.scala:60)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1147)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1147)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
[info]   at scala.collection.immutable.List.foreach(List.scala:392)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
[info]   at scala.collection.immutable.List.foreach(List.scala:392)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
[info]   at org.scalatest.WordSpecLike$class.runTests(WordSpecLike.scala:1147)
[info]   at akka.testkit.AkkaSpec.runTests(AkkaSpec.scala:60)
[info]   at org.scalatest.Suite$class.run(Suite.scala:1147)
[info]   at akka.testkit.AkkaSpec.org$scalatest$WordSpecLike$$super$run(AkkaSpec.scala:60)
[info]   at org.scalatest.WordSpecLike$$anonfun$run$1.apply(WordSpecLike.scala:1192)
[info]   at org.scalatest.WordSpecLike$$anonfun$run$1.apply(WordSpecLike.scala:1192)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
[info]   at org.scalatest.WordSpecLike$class.run(WordSpecLike.scala:1192)
[info]   at akka.testkit.AkkaSpec.org$scalatest$BeforeAndAfterAll$$super$run(AkkaSpec.scala:60)
[info]   at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
[info]   at akka.testkit.AkkaSpec.run(AkkaSpec.scala:60)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:314)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:480)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:301)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:291)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]   at java.lang.Thread.run(Thread.java:748)
[info] - must after receiving Status.Success, allow for earlier completion with PoisonPill (33 milliseconds)
[info] - must fail the stream when receiving Status.Failure (34 milliseconds)
[info] - must set actor name equal to stage name (48 milliseconds)
Coroner Thread Count started at 14, ended at 12, peaked at 21 in akka.stream.scaladsl.ActorRefSourceSpec
[info] Compiling 42 Scala sources to /usr/src/akka-remote-tests/target/multi-jvm-classes ...
[warn] there were 11 deprecation warnings; re-run with -deprecation for details
[warn] one warning found
[info] Done compiling.
[info] Formatting 51 Scala sources ProjectRef(uri("file:/usr/src/"), "akka-cluster")(multi-jvm) ...
[info] Formatting 1 Scala source ProjectRef(uri("file:/usr/src/"), "akka-contrib")(multi-jvm) ...
[info] Formatting 3 Scala sources ProjectRef(uri("file:/usr/src/"), "akka-cluster-typed")(multi-jvm) ...
[info] ScalaTest
[info] Run completed in 19 seconds, 285 milliseconds.
[info] Total number of tests run: 13
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 12, failed 1, canceled 0, ignored 0, pending 0
[info] *** 1 TEST FAILED ***
[error] Failed: Total 13, Failed 1, Errors 0, Passed 12
[error] Failed tests:
[error] 	akka.stream.scaladsl.ActorRefSourceSpec
```

giftig commented:
I've updated the commit message to be a bit more descriptive and follow your contributing guidelines and added a second test to check the `PoisonPill` scenario since I also mentioned that in the original issue.

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/ActorRefSourceSpec.scala:
```diff
@@ -129,6 +130,22 @@ class ActorRefSourceSpec extends StreamSpec {
+      val (ref, done) = {
+        Source.actorRef(3, OverflowStrategy.dropBuffer).toMat(Sink.ignore)(Keep.both).run()
+      }
+      ref ! Status.Success("ok")
```
This can be solved by changing this line: https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/impl/ActorRefSourceActor.scala#L49

to
```
if (bufferSize == 0 || buffer.isEmpty) onCompleteThenStop() // will complete the stream successfully
```

There is another context.stop in there also that I guess should also be changed.

giftig replied:
Good call; that fixed it indeed. I've swapped both of the out of place `context.stop`s with `onCompleteThenStop` and marked the poison pill test as ignore for the moment.

Should the documentation for `Source.actorRef` be updated here: https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala#L473 to acknowledge the bug?

giftig replied:
Perhaps as an alternative to `PoisonPill` some other message (or message-matching function, like the `completionMatcher` function) should trigger the described behaviour (completing without first clearing the buffer). Then a caveat could be added explaining that simply `PoisonPill`ing the actor will not propagate a completion to the stream and should be avoided.

I suspect this probably worked at some point until some changes to the subscriber logic, though (which explains the left over `context.stop`s and the idea that killing the actor would inform the subscriber), so that's potentially an API change.

patriknw replied:
It was probably a bad idea to try to support PoisonPill in the first place. We could add a warning that it doesn't work reliably and should be avoided. Instead use Success.

I doubt completing without draining is a feature that has been asked for, it's probably just a consequence of how PoisonPill works, so we don't have to replace that feature.

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/ActorRefSourceSpec.scala:
```diff
@@ -129,6 +130,22 @@ class ActorRefSourceSpec extends StreamSpec {
+      val (ref, done) = {
+        Source.actorRef(3, OverflowStrategy.dropBuffer).toMat(Sink.ignore)(Keep.both).run()
+      }
+      ref ! PoisonPill
```
This one is trickier because the actor is stopped by the PoisonPill before the subscription arrives so the aroundPostStop can't do much here https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/actor/ActorPublisher.scala#L382

The ActorRefSourceActor is implemented with the deprecated ActorPublisher so I think we should try to rewrite that as a GraphStage instead. I suggest fixing the Success case in this PR and leave this test as pending. Then we'll create a new ticket for the rewrite.

johanandren reviewed (approved):
LGTM but please amend the docs with the mentioned warning about avoiding PoisonPill

giftig commented:
Amended the documentation to explain the perils of `PoisonPill`. One thought: if `PoisonPill` shouldn't be expected to complete the stream after all, it probably makes sense to remove not only my new ignored test but also the existing tests in `ActorRefSourceSpec` which test the behaviour with `PoisonPill`. Does that make sense?

johanandren commented:
Thanks. I agree it makes sense to remove those tests.

giftig commented:
Removed those tests.

raboof reviewed (approved):
Thanks, surprising we did not hit this before!

raboof commented on akka-stream/src/main/scala/akka/stream/impl/ActorRefSourceActor.scala:
```diff
@@ -46,7 +46,7 @@ import akka.stream.ActorMaterializerSettings
 
   def receiveComplete: Receive = completionMatcher.andThen { _ ⇒
-    if (bufferSize == 0 || buffer.isEmpty) context.stop(self) // will complete the stream successfully
+    if (bufferSize == 0 || buffer.isEmpty) onCompleteThenStop() // will complete the stream successfully
```
Can you explain why `ActorPublisher.aroundPostStop` did not take care of this?

giftig replied:
This is my first dive into the internals of akka so I'm not sure I can answer this one. Nearest I can tell the only thing `onCompleteThenStop` does particularly differently is it handles the `PreSubscriber` state as well as the `Active` state; this is something @patriknw was alluding to above I think, but without understanding the wider context of this subscription model I can't comment on why it might be in a pre-subscriber state at the point in which I'm sending the Success in my test; indeed I would have expected these internal subscriptions to all be established once the stream has started.

giftig replied:
I've just been playing around with this in ammonite and it completes ok when I play around actively (with several seconds until I try to complete) so presumably the answer is that it's a race and the stream will fail to complete if it's still in the process of setting up the subscriber, which may still be incomplete when the ActorRef materialises. My test consistently fails without the fix, but that's simply probably because sending a couple of messages right after the actor materializes is consistently faster than establishing that subscription.

patriknw replied:
It's a race condition when sending the PoisonPill before everything is up and running completely

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala:
```diff
@@ -470,13 +470,18 @@ object Source {
+   * failure, will prevent the actor triggering downstream completion and the stream will continue
+   * to run even though the source actor is dead. Therefore you should **not** attempt to
+   * manually terminate the actor such as with a [[akka.actor.PoisonPill]].
+   *
```
also corresponding javadsl/Source

patriknw replied:
There is also in the reference documentation: https://doc.akka.io/docs/akka/current/stream/stream-integrations.html#source-actorref and https://doc.akka.io/docs/akka/current/stream/operators/Source/actorRef.html

giftig replied:
Updated the docs in those places. I also made a couple of grammar fixes and removed some trailing whitespace while I was there.

patriknw reviewed (commented):
a few more things about the docs

patriknw commented:
Thanks!
